File: v2/orchestrator/skills/holidays.py
"""v2 holiday adapters backed by Nager.Date public holidays."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
import re
from typing import Any

import httpx

from lokidoki.core.skill_executor import MechanismResult
from v2.orchestrator.skills._runner import AdapterResult, run_mechanisms

API = "https://date.nager.at/api/v3/PublicHolidays/{year}/{country}"

_COUNTRY_ALIASES = {
    "us": "US",
    "usa": "US",
    "united states": "US",
    "america": "US",
    "japan": "JP",
    "jp": "JP",
    "australia": "AU",
    "au": "AU",
    "canada": "CA",
    "ca": "CA",
    "uk": "GB",
    "united kingdom": "GB",
    "britain": "GB",
    "england": "GB",
}

_HOLIDAY_PREFIXES = (
    "what day is ",
    "when is ",
    "what date is ",
    "what day does ",
    "when does ",
)


@dataclass(slots=True)
class _NagerDateSkill:
    async def execute_mechanism(self, method: str, parameters: dict[str, Any]) -> MechanismResult:
        if method != "nager_public_holidays":
            raise ValueError(f"Unknown mechanism: {method}")
        country = str(parameters.get("country") or "US").upper()
        year = int(parameters.get("year") or date.today().year)
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(
                    API.format(year=year, country=country),
                    headers={"User-Agent": "LokiDoki/0.2"},
                )
        except httpx.HTTPError as exc:
            return MechanismResult(success=False, error=f"network error: {exc}")
        if response.status_code != 200:
            return MechanismResult(success=False, error=f"http {response.status_code}")
        try:
            payload = response.json()
        except ValueError as exc:
            return MechanismResult(success=False, error=f"bad json: {exc}")
        holidays = [
            {
                "name": item.get("localName") or item.get("name") or "",
                "date": item.get("date") or "",
                "global": bool(item.get("global", True)),
            }
            for item in payload
            if isinstance(item, dict)
        ]
        return MechanismResult(
            success=True,
            data={"country": country, "year": year, "holidays": holidays},
            source_url=API.format(year=year, country=country),
            source_title=f"Nager.Date {country} holidays {year}",
        )


_SKILL = _NagerDateSkill()


def _extract_country(text: str, explicit: dict[str, Any]) -> str:
    value = explicit.get("country")
    if value:
        return _COUNTRY_ALIASES.get(str(value).lower().strip(), str(value).upper().strip())
    lower = text.lower()
    for alias, code in sorted(_COUNTRY_ALIASES.items(), key=lambda item: len(item[0]), reverse=True):
        if alias in lower:
            return code
    return "US"


def _extract_year(text: str, explicit: dict[str, Any]) -> int:
    if explicit.get("year"):
        try:
            return int(explicit["year"])
        except (TypeError, ValueError):
            pass
    lower = text.lower()
    today = date.today()
    if "next year" in lower:
        return today.year + 1
    match = re.search(r"\b(20\d{2})\b", lower)
    if match:
        return int(match.group(1))
    return today.year


def _normalize_holiday_name(raw: str) -> str:
    value = raw.lower().strip(" ?.!,'\"")
    for prefix in _HOLIDAY_PREFIXES:
        if value.startswith(prefix):
            value = value[len(prefix):].strip()
            break
    for suffix in (" this year", " next year", " in the us", " in usa", " in japan", " in australia"):
        if value.endswith(suffix):
            value = value[: -len(suffix)].strip()
    return value


def _match_holiday(holidays: list[dict[str, Any]], requested_name: str) -> dict[str, Any] | None:
    query = _normalize_holiday_name(requested_name)
    if not query:
        return None
    compact_query = re.sub(r"[^a-z0-9]+", "", query)
    for item in holidays:
        name = str(item.get("name") or "")
        compact_name = re.sub(r"[^a-z0-9]+", "", name.lower())
        if compact_query == compact_name or compact_query in compact_name or compact_name in compact_query:
            return item
    return None


async def _fetch(country: str, year: int) -> AdapterResult:
    return await run_mechanisms(
        _SKILL,
        [("nager_public_holidays", {"country": country, "year": year})],
        on_success=lambda result, method: "",
        on_all_failed="I couldn't look up holidays right now.",
    )


async def get_holiday(payload: dict[str, Any]) -> dict[str, Any]:
    explicit = payload.get("params") or {}
    text = str(payload.get("chunk_text") or "")
    country = _extract_country(text, explicit)
    year = _extract_year(text, explicit)
    holiday_name = str(explicit.get("name") or text)
    result = await _fetch(country, year)
    if not result.success:
        return result.to_payload()
    holidays = (result.data or {}).get("holidays") or []
    holiday = _match_holiday(holidays, holiday_name)
    if holiday is None:
        return AdapterResult(
            output_text=f"I couldn't find that holiday in {country} for {year}.",
            success=False,
            source_url=result.source_url,
            source_title=result.source_title,
            data=result.data,
            error="holiday not found",
        ).to_payload()
    return AdapterResult(
        output_text=f"{holiday['name']} is on {holiday['date']} in {country} for {year}.",
        success=True,
        mechanism_used="nager_public_holidays",
        source_url=result.source_url,
        source_title=result.source_title,
        data={"holiday": holiday, **(result.data or {})},
    ).to_payload()


async def list_holidays(payload: dict[str, Any]) -> dict[str, Any]:
    explicit = payload.get("params") or {}
    text = str(payload.get("chunk_text") or "")
    country = _extract_country(text, explicit)
    year = _extract_year(text, explicit)
    result = await _fetch(country, year)
    if not result.success:
        return result.to_payload()
    holidays = (result.data or {}).get("holidays") or []
    if not holidays:
        return AdapterResult(
            output_text=f"I couldn't find any holidays for {country} in {year}.",
            success=False,
            source_url=result.source_url,
            source_title=result.source_title,
            data=result.data,
            error="no holidays found",
        ).to_payload()
    preview = ", ".join(f"{item['name']} ({item['date']})" for item in holidays[:5])
    extra = len(holidays) - min(len(holidays), 5)
    suffix = f" Plus {extra} more." if extra > 0 else ""
    return AdapterResult(
        output_text=f"Holidays in {country} for {year}: {preview}.{suffix}",
        success=True,
        mechanism_used="nager_public_holidays",
        source_url=result.source_url,
        source_title=result.source_title,
        data=result.data,
    ).to_payload()


__all__ = [
    "get_holiday",
    "list_holidays",
    "_extract_country",
    "_extract_year",
    "_match_holiday",
]

File: backend/src/tools/holidays.ts
import type { Tool, ToolResult } from './index'

// Hybrid holidays tool: an offline fixed-date table answers common US holidays with no
// network; Nager.Date (https://date.nager.at, no API key) enriches with global coverage,
// movable holidays, and full per-country lists when online. Ported from maipai-home v1/v2.

const NAGER_API = 'https://date.nager.at/api/v3/PublicHolidays'

const COUNTRY_ALIASES: Record<string, string> = {
  us: 'US', usa: 'US', 'united states': 'US', america: 'US', american: 'US',
  uk: 'GB', 'united kingdom': 'GB', britain: 'GB', british: 'GB', england: 'GB',
  canada: 'CA', ca: 'CA', canadian: 'CA',
  australia: 'AU', au: 'AU', australian: 'AU',
  japan: 'JP', jp: 'JP', japanese: 'JP',
  germany: 'DE', de: 'DE', german: 'DE',
  france: 'FR', fr: 'FR', french: 'FR',
  mexico: 'MX', mx: 'MX', mexican: 'MX',
  india: 'IN', in: 'IN', indian: 'IN',
  ireland: 'IE', ie: 'IE', irish: 'IE',
  italy: 'IT', it: 'IT', italian: 'IT',
  spain: 'ES', es: 'ES', spanish: 'ES',
  brazil: 'BR', br: 'BR', brazilian: 'BR',
}

// US fixed-date holidays — instant, offline. Movable ones (Thanksgiving, Easter, MLK Day…)
// are intentionally absent: they require the online table.
const FIXED_US_HOLIDAYS: Record<string, [number, number]> = {
  "new year's day": [1, 1], 'new years day': [1, 1], 'new year': [1, 1],
  "valentine's day": [2, 14], 'valentines day': [2, 14],
  "st patrick's day": [3, 17], "saint patrick's day": [3, 17],
  juneteenth: [6, 19],
  'independence day': [7, 4], 'july fourth': [7, 4], 'fourth of july': [7, 4], '4th of july': [7, 4],
  halloween: [10, 31],
  'veterans day': [11, 11], "veteran's day": [11, 11],
  'christmas eve': [12, 24],
  christmas: [12, 25], 'christmas day': [12, 25], xmas: [12, 25],
  "new year's eve": [12, 31], 'new years eve': [12, 31],
}

const HOLIDAY_PREFIXES = [
  'what day is the', 'what date is the', 'what day does', 'when does',
  'what day is', 'what date is', 'when is the', 'when is', 'when',
  'how many days until', 'days until',
]
const HOLIDAY_SUFFIXES = [
  ' this year', ' next year', ' last year',
  ' in the us', ' in usa', ' in the usa', ' in america',
  ' fall on', ' fall', ' happen', ' celebrated', ' observed',
]

const WEEKDAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
const MONTHS = ['January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December']

function compact(s: string): string {
  return s.toLowerCase().replace(/[^a-z0-9]+/g, '')
}

function detectCountry(text: string, config?: Record<string, unknown>): string {
  const lower = text.toLowerCase()
  // longest alias first so "united states" wins over "us"
  for (const alias of Object.keys(COUNTRY_ALIASES).sort((a, b) => b.length - a.length)) {
    if (new RegExp(`\\b${alias.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`).test(lower)) {
      return COUNTRY_ALIASES[alias]!
    }
  }
  const cfg = String(config?.country ?? '').trim()
  if (cfg) return COUNTRY_ALIASES[cfg.toLowerCase()] ?? cfg.toUpperCase()
  return 'US'
}

function detectYear(text: string): number {
  const now = new Date()
  const lower = text.toLowerCase()
  if (lower.includes('next year')) return now.getFullYear() + 1
  if (lower.includes('last year')) return now.getFullYear() - 1
  const m = lower.match(/\b(20\d{2})\b/)
  if (m) return parseInt(m[1]!, 10)
  return now.getFullYear()
}

function isListRequest(text: string): boolean {
  const lower = text.toLowerCase()
  if (/\b(list|all)\b/.test(lower) && lower.includes('holiday')) return true
  if (/\b(what|which|any)\b[^?]*\bholidays\b/.test(lower)) return true
  if (/\bholidays\b[^?]*\b(in|for|during)\b/.test(lower)) return true
  return false
}

function normalizeHolidayName(raw: string): string {
  let v = raw.toLowerCase().trim().replace(/^[\s?.!,'"]+|[\s?.!,'"]+$/g, '')
  for (const p of HOLIDAY_PREFIXES) {
    if (v.startsWith(p + ' ')) { v = v.slice(p.length).trim(); break }
  }
  for (const s of HOLIDAY_SUFFIXES) {
    if (v.endsWith(s)) { v = v.slice(0, -s.length).trim(); break }
  }
  // drop a trailing country phrase, e.g. "christmas in japan"
  v = v.replace(/\b(in|for)\s+[a-z .]+$/i, '').trim()
  return v
}

function formatDate(iso: string): { weekday: string; pretty: string } {
  const parts = iso.split('-').map(Number)
  const y = parts[0] ?? 1970, m = parts[1] ?? 1, d = parts[2] ?? 1
  const dt = new Date(Date.UTC(y, m - 1, d))
  const wd = WEEKDAYS[dt.getUTCDay()] ?? ''
  return {
    weekday: wd,
    pretty: `${wd}, ${MONTHS[m - 1] ?? ''} ${d}, ${y}`,
  }
}

interface NagerHoliday { name: string; date: string; global: boolean }

async function fetchNager(country: string, year: number): Promise<{ holidays: NagerHoliday[]; url: string } | null> {
  const url = `${NAGER_API}/${year}/${country}`
  const res = await fetch(url, {
    headers: { Accept: 'application/json', 'User-Agent': 'MaiPaiHome/3' },
    signal: AbortSignal.timeout(5000),
  })
  if (!res.ok) return null
  const payload = await res.json() as Array<Record<string, unknown>>
  if (!Array.isArray(payload)) return null
  return {
    url,
    holidays: payload.map(item => ({
      name: String(item.localName || item.name || ''),
      date: String(item.date || ''),
      global: Boolean(item.global ?? true),
    })).filter(h => h.name && h.date),
  }
}

function matchHoliday(holidays: NagerHoliday[], requested: string): NagerHoliday | null {
  const q = compact(normalizeHolidayName(requested))
  if (!q) return null
  // Exact match wins outright; otherwise fall back to the first substring match.
  let partial: NagerHoliday | null = null
  for (const h of holidays) {
    const n = compact(h.name)
    if (n === q) return h
    if (!partial && (n.includes(q) || q.includes(n))) partial = h
  }
  return partial
}

function offlineLookup(holidayName: string, year: number): { name: string; date: string } | null {
  const q = compact(normalizeHolidayName(holidayName))
  if (!q) return null
  const build = (name: string, month: number, day: number) => ({
    // Capitalize each word's first letter only — leaves "year's" intact.
    name: name.replace(/(^|\s)\w/g, c => c.toUpperCase()),
    date: `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`,
  })
  let partial: { name: string; date: string } | null = null
  for (const [name, [month, day]] of Object.entries(FIXED_US_HOLIDAYS)) {
    const n = compact(name)
    if (n === q) return build(name, month, day)
    if (!partial && (n.includes(q) || q.includes(n))) partial = build(name, month, day)
  }
  return partial
}

export const holidaysTool: Tool = {
  id: 'holidays',
  name: 'Holidays',
  description: 'Look up when a holiday falls, or list public holidays for a country and year (offline for common US holidays, global via Nager.Date when online)',
  offline: true,
  dataSources: [
    { name: 'Nager.Date', domain: 'date.nager.at', purpose: 'Public holiday dates for countries worldwide (used when online)', type: 'api' },
  ],
  passMessage: 'query',
  configSchema: [
    {
      key: 'country',
      label: 'Default Country',
      description: 'ISO country code used when no country is mentioned (e.g. US, GB, JP)',
      type: 'string' as const,
      scope: 'user' as const,
      default: 'US',
      placeholder: 'US',
    },
  ],
  examples: [
    'when is a specific holiday this year or next year',
    'what day does a holiday fall on',
    'list the public holidays for a country and year',
    'what holidays are coming up in a given country',
    'the date of a national or public holiday',
  ],
  toolDefinition: {
    type: 'function',
    function: {
      name: 'holidays',
      description: 'Look up a holiday date or list public holidays for a country/year',
      parameters: {
        type: 'object',
        required: ['query'],
        properties: {
          query: { type: 'string', description: 'The holiday question, e.g. "when is Christmas" or "holidays in Japan next year"' },
        },
      },
    },
  },

  async execute(args: unknown, config?: Record<string, unknown>): Promise<ToolResult> {
    const a = (args ?? {}) as { query?: string }
    const text = (a.query ?? '').trim()
    if (!text) return { success: false, error: 'A holiday question is required' }

    const country = detectCountry(text, config)
    const year = detectYear(text)
    const listMode = isListRequest(text)

    // Fast offline path: a specific common US holiday with a fixed date.
    if (!listMode && country === 'US') {
      const hit = offlineLookup(text, year)
      if (hit) {
        const { weekday, pretty } = formatDate(hit.date)
        // Best-effort online enrich (more accurate localName), but never block on it.
        return {
          success: true,
          data: {
            mode: 'lookup', found: true, country, year,
            holiday: { name: hit.name, date: hit.date, weekday },
            answer_payload: {
              gist: `${hit.name} ${year} falls on ${pretty}.`,
              highlights: [], sources: [], depth_available: false,
            },
          },
        }
      }
    }

    // Online path (Nager.Date) for movable holidays, other countries, and full lists.
    let nager: Awaited<ReturnType<typeof fetchNager>> = null
    try {
      nager = await fetchNager(country, year)
    } catch {
      nager = null
    }

    if (!nager) {
      // Degrade gracefully to the offline table.
      if (listMode) {
        const seen = new Set<string>()
        const list = Object.entries(FIXED_US_HOLIDAYS)
          .map(([name, [m, d]]) => ({ name: name.replace(/(^|\s)\w/g, c => c.toUpperCase()), date: `${year}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}` }))
          .filter(h => { const k = h.date; if (seen.has(k)) return false; seen.add(k); return true })
          .sort((x, y) => x.date.localeCompare(y.date))
        return {
          success: true, offline: true,
          data: {
            mode: 'list', found: true, country: 'US', year, partial: true, holidays: list,
            answer_payload: {
              gist: `I'm offline, so here are common US holidays for ${year} (fixed dates only):`,
              highlights: list.map(h => `${h.name}: ${formatDate(h.date).pretty}`),
              sources: [], depth_available: false,
            },
          },
        }
      }
      const off = offlineLookup(text, year)
      if (off) {
        const { weekday, pretty } = formatDate(off.date)
        return {
          success: true, offline: true,
          data: {
            mode: 'lookup', found: true, country: 'US', year,
            holiday: { name: off.name, date: off.date, weekday },
            answer_payload: { gist: `${off.name} ${year} falls on ${pretty}.`, highlights: [], sources: [], depth_available: false },
          },
        }
      }
      return { success: false, offline: true, error: 'Holiday lookup needs an internet connection for that request.' }
    }

    const source = { url: nager.url, title: `Public holidays — ${country} ${year}` }

    if (listMode) {
      const sorted = [...nager.holidays].sort((x, y) => x.date.localeCompare(y.date))
      return {
        success: true,
        data: {
          mode: 'list', found: sorted.length > 0, country, year, holidays: sorted.map(h => ({ name: h.name, date: h.date, weekday: formatDate(h.date).weekday })),
          answer_payload: {
            gist: sorted.length ? `There are ${sorted.length} public holidays in ${country} for ${year}.` : `No public holidays found for ${country} in ${year}.`,
            highlights: sorted.slice(0, 12).map(h => `${h.name}: ${formatDate(h.date).pretty}`),
            sources: [source], depth_available: sorted.length > 12,
          },
        },
      }
    }

    const match = matchHoliday(nager.holidays, text)
    if (!match) {
      return {
        success: true,
        data: {
          mode: 'lookup', found: false, country, year,
          answer_payload: { gist: `I couldn't find that holiday in ${country} for ${year}.`, highlights: [], sources: [source], depth_available: false },
        },
      }
    }
    const { weekday, pretty } = formatDate(match.date)
    return {
      success: true,
      data: {
        mode: 'lookup', found: true, country, year,
        holiday: { name: match.name, date: match.date, weekday },
        answer_payload: { gist: `${match.name} in ${country} ${year} falls on ${pretty}.`, highlights: [], sources: [source], depth_available: false },
      },
    }
  },
}
